lib/subscribers/message-consumer.js

```diff
@@ -66,11 +66,21 @@ class MessageConsumerSubscriber extends Subscriber {
   }
 
   /**
-   * Ends the transaction created for the consumption callback.
+   * Checks the result of the message handler.
+   * If it's a promise, waits for it to resolve before ending transaction.
+   * This ensures that the transaction stays active until the promise resolves.
+   * @param {object} data the data associated with the `asyncEnd` event
    */
-  asyncEnd() {
+  asyncEnd(data) {
     const ctx = this.agent.tracer.getContext()
-    ctx?.transaction?.end()
+    const result = data.cbResult
+    if (result instanceof Promise) {
```

Comment:
It could be a thenable or even some random non-native `Promise` implementation. We should instead:

```suggestion
    if (typeof result.then === 'function') {
```

Reply:
true. I'll update